py/server/deephaven/barrage.py

```diff
@@ -15,23 +15,88 @@
 from deephaven.table import Table
 
 _JURI = jpy.get_type("java.net.URI")
+_JTimeUnit = jpy.get_type("java.util.concurrent.TimeUnit")
 _JClientConfig = jpy.get_type("io.deephaven.client.impl.ClientConfig")
 _JSSLConfig = jpy.get_type("io.deephaven.ssl.config.SSLConfig")
 _JSessionImplConfig = jpy.get_type("io.deephaven.client.impl.SessionImplConfig")
 _JTrustCustom = jpy.get_type("io.deephaven.ssl.config.TrustCustom")
 _JDeephavenTarget = jpy.get_type("io.deephaven.uri.DeephavenTarget")
 _JBarrageSession = jpy.get_type("io.deephaven.client.impl.BarrageSession")
-_JSharedId = jpy.get_type("io.deephaven.client.impl.SharedId")
+_JTableSpec = jpy.get_type("io.deephaven.qst.table.TableSpec")
 _JBarrageTableResolver = jpy.get_type("io.deephaven.server.uri.BarrageTableResolver")
 _JChannelHelper = jpy.get_type("io.deephaven.client.impl.ChannelHelper")
 _JDeephavenChannelImpl = jpy.get_type("io.deephaven.proto.DeephavenChannelImpl")
 _JSessionImpl = jpy.get_type("io.deephaven.client.impl.SessionImpl")
 _JExecutors = jpy.get_type("java.util.concurrent.Executors")
 _JRootAllocator = jpy.get_type("org.apache.arrow.memory.RootAllocator")
 
-_session_cache: Dict[str, BarrageSession] = {}  # use WeakValueDictionary to avoid memory leak?
-_remote_session_lock = threading.Lock()
+class BarrageSession():
+    """ A Deephaven Barrage session to a remote server."""
+
+    def __init__(self, j_barrage_session: jpy.JType, j_managed_channel: jpy.JType = None):
+        self.j_barrage_session = j_barrage_session
+        self.j_session = j_barrage_session.session()
+        self.j_managed_channel = j_managed_channel
+
+    def __enter__(self) -> BarrageSession:
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.close()
+
+    def close(self) -> None:
+        """ Closes the session."""
+        try:
+            self.j_barrage_session.close()
+            if self.j_managed_channel:
+                self.j_managed_channel.shutdown()
+                while not self.j_managed_channel.awaitTermination(1, _JTimeUnit.MILLISECONDS):
+                    pass
```

Comment:
Why the tight loop every millisecond? Seems a bit too frequent. Also, we'll be waiting forever if there's some sort of issue in awaiting termination.